### PR TITLE
Input Stream API change

### PR DESCRIPTION
**Pull request by TingDaoK**
API change:
- Refcount `aws_input_stream`
- API change, every down stream makes its own input stream implementation needs to be updated.
- Introduced `release/acquire` vtable function for language bindings that need to use their own refcounts.

### will not merge until we fixed the downstream that creates its own inputstream

C:
- https://github.com/awslabs/aws-c-s3/pull/175
- https://github.com/awslabs/aws-c-http/pull/342
- https://github.com/awslabs/aws-c-auth/pull/159

CRTs:
- https://github.com/awslabs/aws-crt-python/pull/336
- https://github.com/awslabs/aws-crt-nodejs/pull/301
- https://github.com/awslabs/aws-crt-java/pull/447
- https://github.com/awslabs/aws-crt-cpp/pull/341
- https://github.com/awslabs/aws-crt-dotnet/pull/81
- https://github.com/awslabs/aws-crt-ffi/pull/40

**For aws-crt-*, we will need to update them when they pull the latest submodules**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/io/stream.h
Comment (graebm) [inline include/aws/io/stream.h:48]: debatable/trivial: instead of passing in `void *impl`, what if we passed in `size_t sizeof_impl` and then new() did `aws_mem_acquire_many()` to create both the base-struct and impl in one go?

File: source/stream.c
Comment (graebm) [inline source/stream.c:16]: trivial: const `const struct aws_input_stream_vtable *vtable`

File: source/stream.c
Comment (graebm) [inline source/stream.c:88]: allocations cannot fail. remove this check

File: source/stream.c
Comment (graebm) [inline source/stream.c:241]: remove check for allocation failure

File: source/stream.c
Comment (graebm) [inline source/stream.c:332]: remove OOM check

File: source/stream.c
Comment (graebm) [inline source/stream.c:359]: remove OOM check

File: source/stream.c
Comment (graebm) [reply source/stream.c]: oops. this comment is from weeks ago...

File: include/aws/io/stream.h
Comment (TingDaoK) [reply include/aws/io/stream.h:48]: hmmm, I feel it is a bit weird as the `impl` will still need a callback to destroy (To clean up resource besides of the memory allocation). So, the destroy callback will not need to clean up the memory seems like a bit weird to me.

**Review by TwistedTwigleg — COMMENTED** [review]
Looks good to me!
There was only one minor thing I noticed, and it's more of a question than anything.

Because I'm not experienced with this code yet, I would wait for another review just to make sure it is all good, as it's entirely possible I missed something.

File: source/stream.c
Comment (TingDaoK) [reply source/stream.c:69]: yeah, good catch. It should be static

Comment (TingDaoK) [reply source/stream.c]: Well, what other function starts with `s_` but not being static? I didn't notice any from a quick search

Comment (TwistedTwigleg) [reply source/stream.c]: I just looked and now I cannot find it. I thought it was in “stream.c” or maybe the transport handler, but I looked again just now in all the changed files and didn’t see any.

I’ll take another look in the morning though to be sure, but maybe I just thought I saw multiple 🤔

Comment (TwistedTwigleg) [reply source/stream.c]: I looked again and realized what it was: I was the static `struct` creations that were throwing me off, as thought the class inheritance was the name. For example, I thought `aws_channel_handler_vtable` was the name of the struct in `static struct aws_channel_handler_vtable s_handler_vtable`, not `s_handler_vtable`.

Apologizes for the confusion! I looked through all of the files and can confirm all the statics now start with `s_`.

**Review by bretambrose — COMMENTED** [review]
This pattern is a bit different than what we've done in the past and I'm unsure that there's a need for that difference.

What I'd prefer to see:

aws_input_stream stays public (remove allocator, add ref count)
aws_input_stream is embedded as a value within the actual implementations, can remove the get_impl function as well
destroy is removed from the vtable because the refcount-to-zero callback -- set at subclass creation time -- will do the full cleanup.

File: source/stream.c
Comment (graebm) [inline source/stream.c:68]: trivial: usually we make `base` is the very first member

**Review by graebm — APPROVED** [review]

File: include/aws/io/stream.h
Comment (graebm) [inline include/aws/io/stream.h:52]: this doesn't need to return anything
you can do that from aws_input_stream_acquire()

File: include/aws/io/stream.h
Comment (graebm) [inline include/aws/io/stream.h:49]: This sentence doesn't make a ton of sense, just remove it. previous sentence says it nicely

Comment (graebm) [reply include/aws/io/stream.h:49]: Add comment about how the refcount isn't used if you override acquire()/release()

**Review by graebm — APPROVED** [review]
fix & ship

File: include/aws/io/stream.h
Comment (bretambrose) [reply include/aws/io/stream.h:52]: This is something I grumped about earlier and twisted Dengke's arm to include.  I prefer the return value of always-null to simplify non-final cleanup (just assign the result) and swaps

Comment (TingDaoK) [reply include/aws/io/stream.h:52]: yeah, we agreed that the c function should return the pointer, but the implementation doesn't need to.

**Pull request merged**
